packages/amplify-appsync-simulator/src/velocity/value-mapper/map.ts

```diff
@@ -56,7 +57,7 @@ export class JavaMap {
   }
 
   keySet() {
-    return new JavaArray(Array.from(this.map.keys()), this.mapper);
+    return new JavaArray(Array.from(this.map.keys()).map(toJavaString), toJavaString);
```

Comment:
Is this new `toJavaString()` function needed. Shouldn't `this.mapper` handle the conversion [here](https://github.com/aws-amplify/amplify-cli/blob/5c9172af5dbd911ce86823f32e774e4c89fd9f88/packages/amplify-appsync-simulator/src/velocity/value-mapper/mapper.ts#L27-L30)?

Reply:
From what I see and the tests that I've run, for the array values that are passed to the constructor, the mapper isn't called.

Reply:
Sorry, let me rephrase my question. For this piece specifically:

```js
Array.from(this.map.keys()).map(toJavaString)
```

You are getting all of the keys and then mapping them to Java strings using the new `toJavaString()` function. That `map()` call is new. Instead of creating the new `toJavaString()` function, couldn't you pass the existing mapper there, which should handle the conversion via the code I linked to above?

Reply:
ah, I see your point! 

Let me make that change. Thanks for pointing that out

Reply:
@cjihrig,

In the interest of avoiding further back and forth. 

Is this what you mean? 
```
keySet() {
    return this.mapper(Array.from(this.map.keys()));
  }
```

Reply:
@cjihrig,

You good if we invoke `map` directly instead of using `this.mapper`, since `this.mapper` may or may not contain logic to convert a `string` to `JavaString`.

```
  keySet() {
    return map(Array.from(this.map.keys()));
  }
```

To add a bit more context, [this test](https://github.com/aws-amplify/amplify-cli/blob/5c9172af5dbd911ce86823f32e774e4c89fd9f88/packages/amplify-appsync-simulator/src/__tests__/velocity/value-mapper/map.test.ts#L65) will break if we use `this.mapper`

Apologies in case I misunderstood your suggestion.

Reply:
The test suite passed for me with no modifications with this line:

```js
return new JavaArray(Array.from(this.map.keys()).map(this.mapper as any), this.mapper);
```

Reply:
just to double-check with you though, `this.mapper` may not always resolve a string to JavaString. 

For example, in the tests, we use `identityMapper`  which results in the newly added test, that expects a JavaArray with each element being a JavaString, to fail. 

Would there be a problem if we use this:

```
  keySet() {
    return map(Array.from(this.map.keys()));
  }
```

Reply:
> just to double-check with you though, this.mapper may not always resolve a string to JavaString.

I think that is expected behavior. The mapper is responsible for converting input values to output values. The `identityMapper` just doesn't do any real mapping.

> Would there be a problem if we use this:

I think so. The reason for the `mapper` option is to customize the mapping behavior. Hard coding the mapper like that would break that. For your new test, you could use the real mapper though.

Reply:
The way I was thinking about it, is that the `keySet` function should always return a JavaArray < JavaString >, irrespective of the mapper that is passed to the JavaMap. 

Is there a case where the return value of the keySet would be dependant on the mapper of the JavaMap?

Reply:
Again, thank you so much for your time in reviewing this and patiently answering my questions.